### DISCUSSION TITLE
详解 Golang 中间代码生成 · /golang-ir-ssa

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#545] [Announcements]
https://draveness.me/golang-ir-ssa

**Comment by woofyzhao** [comment]
buildssa 就是用来具有 SSA 特性的中间代码的函数 -> 少了个"生成"?

---

*2020-08-27 UPDATES: 已修复*

**Comment by tdergouzi** [comment]
请教一下，多轮替换的过程中，ssa.Compile 函数会删除打印日志和性能分析的代码，如果我理解为开发者为调试而刻意添加了日志代码或者性能分析相关代码，是否就会有问题？

**Comment by draveness** [comment]
> 请教一下，多轮替换的过程中，ssa.Compile 函数会删除打印日志和性能分析的代码，如果我理解为开发者为调试而刻意添加了日志代码或者性能分析相关代码，是否就会有问题？

我没有太理解你问题的场景，要调试 Go 语言还是要调试用 Go 语言写的服务，你可以试试增加的日志代码会不会删除，个人觉得不会

**Comment by jianfengye** [comment]
请教一下，从ssa码中怎么查找对应的执行逻辑，比如在SSA中看到 SliceMake ，但是怎么找到对应的逻辑代码呢？

**Comment by xing393939** [comment]
"经过将近 50 轮处理的中间代码相比处理之前有了非常大的改变，执行效率会有比较大的提升，多轮的处理已经包含了一些机器特定的修改，包括根据目标架构对代码进行改写"

这里的意思是在生成SSA中间代码的时候就要针对不同cpu架构生成代码了？我一直理解的是编译四阶段，前三个阶段最终生成的SSA中间代码是与cpu结构和os无关的，只有第四阶段用汇编器生成目标代码时才考虑cpu架构和os。

**Reply by LycorisFubuki** [reply]
https://medium.com/a-journey-with-go/go-overview-of-the-compiler-4e5a153ca889 也许对你有帮助

**Comment by TenshGU** [comment]
AST到SSA
cmd/compile/internal/gc.state.stmt 方法贴的源码有误
应该是：
```
case OCALLMETH, OCALLINTER:
		s.callResult(n, callNormal)
```

**Comment by ikingye** [comment]
https://draven.co/golang/tree/cmd/compile/internal/gc.Main 跳转 https://github.com/golang/go/blob/d902791b509b641683d4ec58b282180c56918aec/src/cmd/compile/internal/gc/main.go#L148 是怎么实现的，好维护吗
